utils/bids/bids_file_locator.py

```diff
@@ -0,0 +1,248 @@
+
+                    for dirnum in set(dwi_dict['bval']):
+                        ## the following statement extracts index values in dwi_dict['bval'] if the value matches
+                        # "dirnum", which is the number of directions (i.e. 98 or 99). These index values are used
+                        # to find the corresponding PE directions, dwi file names, etc. in the dictionary
+                        idxs = { i for k,v in dwi_dict.iteritems() for i in range(0,len(dwi_dict['bval'])) if v[i] == dirnum }
+                        PEdirNums = set([dwi_dict['direction'][i] for i in idxs])
+                        for PEdirNum in PEdirNums:
+                            dwis = [ dwi_dict['dwiFile'][i] for i in idxs if dwi_dict['direction'][i] == PEdirNum ]
+                            assert len(dwis) <= 2
+                            dwiname = "Diffusion" + "_dir-" + str(dirnum) + "_" + numrun + "_corr_" + str(PEdirNum)
+                            if "j" in PEdirNum:
+                                PEdir = 2
+                            elif "i" in PEdirNum:
+                                PEdir = 1
+                            else:
+                                RuntimeError("Phase encoding direction not specified for diffusion data.")
+                            pos = "EMPTY"
+                            neg = "EMPTY"
+                            gdcoeffs = "None"
+                            for dwi in dwis:
+                                if "-" in layout.get_metadata(dwi)["PhaseEncodingDirection"]:
+                                    neg = dwi
+                                else:
+                                    pos = dwi
+
+                                echospacing = layout.get_metadata(pos)["EffectiveEchoSpacing"] * 1000
+    #TODO make sure there are not naming collisions in the config dict such that parameters (e.g., echospacing
+    #is overwritten from structural setup to func to diff
+                    self.config.update({
+                        "dwiname":dwiname,
+                        "PEdir":PEdir,
+                        "pos": pos,
+                        "neg":neg,
+                        "echospacing": echospacing
+                    })
+
+    def grab_BIDS_data(self):
+        """
+        Download and validate the BIDS directory structure from curated data.
+        This method must be included since the files are copied up to Docker and the
+        file organization will need to follow the BIDS curation/specification.
+        Args:
+            gtk_context (GearToolkitContext): gear information
+        Returns:
+            msg (str): Status of BIDS download (and validation)
+        """
+        log.info("Checking BIDS")
+        config = self.gtk_context.config
+        try:
+            log.info('Downloading BIDS')
+            self.gtk_context.download_session_bids(target_dir=self.gtk_context.work_dir / "bids")
```

Comment:
Don't you want something like [download_bids_for_runlevel()](https://gitlab.com/flywheel-io/flywheel-apps/templates/bids-app-template/-/blob/main/utils/bids/download_run_level.py) here?  I would think it needs to load all data for a subject (all sessions).

Reply:
Yes - I still have things to learn about the SDK and GearToolkit. Thanks!